### PR TITLE
Kafka Mirror Maker status

### PR DESCRIPTION
**Pull request by stanlyDoge**
### Type of change
- Enhancement / new feature

### Description
Introducing status subresource for the Kafka Mirror Maker.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [ ] Update/write design documentation in `./design`
- [ ] Write tests
- [ ] Make sure all tests pass
- [ ] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [ ] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md

### CODE REVIEW

**Comment by stanlyDoge** [comment]
Ready MM
```
status:
  conditions:
  - lastTransitionTime: 2019-07-11T13:01:54+0000
    status: "True"
    type: Ready
  name: my-mirror-maker
  observedGeneration: 1
```
Unready MM
```
status:
  conditions:
  - lastTransitionTime: 2019-07-11T13:03:04+0000
    message: 'Version 2.2.8 is not supported. Supported versions are: 2.1.0, 2.1.1,
      2.2.0, 2.2.1.'
    reason: InvalidResourceException
    status: "True"
    type: NotReady
  observedGeneration: 2
```

**Comment by scholzj** [comment]
Again, I do not understand what the value of the name would be. I would remove the name and keep observed generation and conditions.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 1]
```diff
@@ -84,13 +94,88 @@ public KafkaMirrorMakerAssemblyOperator(Vertx vertx, PlatformFeaturesAvailabilit
...
+     */
+    Future<Void> updateStatus(KafkaMirrorMaker kafkaMirrorMakerAssembly, Reconciliation reconciliation, KafkaMirrorMakerStatus desiredStatus) {
+        Future<Void> updateStatusFuture = Future.future();
+
+        resourceOperator.getAsync(kafkaMirrorMakerAssembly.getMetadata().getNamespace(), kafkaMirrorMakerAssembly.getMetadata().getName()).setHandler(getRes -> {
+            if (getRes.succeeded()) {
+                KafkaMirrorMaker mirrorMaker = getRes.result();
+
+                if (mirrorMaker != null) {
+                    if ("kafka.strimzi.io/v1alpha1".equals(mirrorMaker.getApiVersion())) {
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: I did comment about this hard-coded check on the KafkaUser status PR.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 2]
```diff
@@ -85,13 +95,88 @@ public KafkaMirrorMakerAssemblyOperator(Vertx vertx, PlatformFeaturesAvailabilit
...
+                                    updateStatusFuture.fail(updateRes.cause());
+                                }
+                            });
+                        } else {
+                            log.debug("{}: Status did not change", reconciliation);
+                            updateStatusFuture.complete();
+                        }
+                    }
+                } else {
+                    log.error("{}: Current Kafka resource not found", reconciliation);
```
Comment (ppatierno) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: Kafka Mirror Maker

File: api/src/main/java/io/strimzi/api/kafka/model/status/KafkaMirrorMakerStatus.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+@Buildable(
+        editableEnabled = false,
+        generateBuilderPackage = false,
+        builderPackage = "io.fabric8.kubernetes.api.builder"
+)
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@JsonPropertyOrder({"conditions", "observedGeneration"})
+@EqualsAndHashCode
+public class KafkaMirrorMakerStatus extends Status {
+    private static final long serialVersionUID = 1L;
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/status/KafkaMirrorMakerStatus.java]: could we add information about which are the source and destination clusters in mirroring?

Comment (scholzj) [reply api/src/main/java/io/strimzi/api/kafka/model/status/KafkaMirrorMakerStatus.java]: How do you think it should look like? Isn't it just mirroring the `spec` section?

Comment (ppatierno) [reply api/src/main/java/io/strimzi/api/kafka/model/status/KafkaMirrorMakerStatus.java]: Yes it is, actually. Maybe it's useless then.

File: cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java [hunk 4]
```diff
@@ -85,13 +95,88 @@ public KafkaMirrorMakerAssemblyOperator(Vertx vertx, PlatformFeaturesAvailabilit
...
+        Future<Void> chainFuture = Future.future();
+        mirrorMakerServiceAccount(namespace, mirror)
                 .compose(i -> deploymentOperations.scaleDown(namespace, mirror.getName(), mirror.getReplicas()))
                 .compose(scale -> serviceOperations.reconcile(namespace, KafkaMirrorMakerResources.serviceName(mirror.getCluster()), mirror.generateService()))
                 .compose(i -> configMapOperations.reconcile(namespace, mirror.getAncillaryConfigName(), logAndMetricsConfigMap))
                 .compose(i -> podDisruptionBudgetOperator.reconcile(namespace, mirror.getName(), mirror.generatePodDisruptionBudget()))
                 .compose(i -> deploymentOperations.reconcile(namespace, mirror.getName(), mirror.generateDeployment(annotations, pfa.isOpenshift(), imagePullPolicy, imagePullSecrets)))
-                .compose(i -> deploymentOperations.scaleUp(namespace, mirror.getName(), mirror.getReplicas()).map((Void) null));
+                .compose(i -> deploymentOperations.scaleUp(namespace, mirror.getName(), mirror.getReplicas()))
+                .compose(i -> deploymentOperations.readiness(namespace, mirror.getName(), 1_000, 30_000))
```
Comment (scholzj) [inline cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: Use the user configured operation timeout.

Comment (stanlyDoge) [reply cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaMirrorMakerAssemblyOperator.java]: By this you mean change introduced in https://github.com/strimzi/strimzi-kafka-operator/pull/1748/files#diff-089830c09c2acee56463cc61fed16fa7R90, right? I need to wait until this change is in the master :(

File: documentation/book/appendix_crds.adoc [hunk 5]
```diff
@@ -1645,9 +1645,11 @@ Used in: xref:type-KafkaUser-{context}[`KafkaUser`]
 
 [options="header"]
 |====
-|Property     |Description
-|spec  1.2+<.<|The specification of the mirror maker.
+|Property       |Description
+|spec    1.2+<.<|The specification of the mirror maker.
```
Comment (laidan6000) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|spec    1.2+<.<|The specification of Kafka MirrorMaker.
```

File: documentation/book/appendix_crds.adoc [hunk 6]
```diff
@@ -1822,6 +1824,21 @@ Used in: xref:type-KafkaMirrorMakerSpec-{context}[`KafkaMirrorMakerSpec`]
...
+
+Used in: xref:type-KafkaMirrorMaker-{context}[`KafkaMirrorMaker`]
+
+
+[options="header"]
+|====
+|Property                   |Description
+|conditions          1.2+<.<|List of status conditions.
+|xref:type-Condition-{context}[`Condition`] array
+|observedGeneration  1.2+<.<|The generation of the CRD which was last reconciled by the operator.
```
Comment (laidan6000) [inline documentation/book/appendix_crds.adoc]:
```suggestion
|observedGeneration  1.2+<.<|The generation of the CRD that was last reconciled by the operator.
```

File: documentation/book/appendix_crds.adoc [hunk 7]
```diff
@@ -1645,9 +1645,11 @@ Used in: xref:type-KafkaUser-{context}[`KafkaUser`]
 
 [options="header"]
 |====
-|Property     |Description
-|spec  1.2+<.<|The specification of the mirror maker.
+|Property       |Description
+|spec    1.2+<.<|The specification of the mirror maker.
 |xref:type-KafkaMirrorMakerSpec-{context}[`KafkaMirrorMakerSpec`]
+|status  1.2+<.<|The status of the Kafka Mirror Maker.
```
[the remark this answers is not in the document]
Comment (scholzj) [reply documentation/book/appendix_crds.adoc]: When you say _status of all Kafka Mirror Maker replicas_ I would expect per replica status. That is not the case, there is just one status for the whole Kafka Mirror Maker cluster.

[the remark this answers is not in the document]
Comment (laidan6000) [reply documentation/book/appendix_crds.adoc]: OK, makes sense. Please use: "The status of Kafka Mirror Maker".

@stanlyDoge

**Comment by stanlyDoge** [comment]
@strimzi-ci run tests

**Pull request merged**